### PR TITLE
A `relay` server written on Rust.

### LINKED ISSUE #3477: [sBTC] Convert `data relay` Python script to Rust
**Is your feature request related to a problem? Please describe.**
We would like to use only few programming languages. Currently, the main language for the repo is Rust. The specific [script](https://github.com/stacks-network/stacks-blockchain/blob/sbtc/contrib/tools/http_datarelay.py) is written in Python.

**Describe the solution you'd like**
We will use a Rust program instead. Also, we will have a Rust library for unit testing.

**Describe alternatives you've considered**
Keep it as a Python script but it will increase a number of languages in the repo. Also it is hard to use the Python script for Rust unit testing. 

**Additional context**

### PR DESCRIPTION
**Pull request by sergey-shandar**
### Applicable issues
- fixes #3477 

### Checklist
- [ ] Test coverage for new or modified code paths
- [ ] Changelog is updated
- [ ] Required documentation changes (e.g., `docs/rpc/openapi.yaml` and `rpc-endpoints.md` for v2 endpoints, `event-dispatcher.md` for new events)
- [ ] New clarity functions have corresponding PR in `clarity-benchmarking` repo
- [ ] New integration test(s) added to `bitcoin-tests.yml`

### CODE REVIEW

File: contrib/tools/relay-server/README.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# A Relay Server
+
+## Start the `relay-server` server
```
Comment (igorsyl) [inline contrib/tools/relay-server/README.md:3]: Can you add a description of this application:
- What does it do?
- What port does it listen to?
- How does it relay messages?

**Comment by xoloki** [comment]
One high level question: why are we implementing HTTP instead of simply using an existing library (e.g. hyper)?

**Review by netrome — APPROVED** [review]

**Review by donpdonp — APPROVED** [review]
LGTM

**Comment by sergey-shandar** [comment]
> One high level question: why are we implementing HTTP instead of simply using an existing library (e.g. hyper)?

We would like to use the code in test and easily mock any IO. However, we may use some utility functions from third-party libraries, for example encode/decode HTTP messages.

**Comment by xoloki** [comment]
> > One high level question: why are we implementing HTTP instead of simply using an existing library (e.g. hyper)?
> 
> We would like to use the code in test and easily mock any IO. However, we may use some utility functions from third-party libraries, for example encode/decode HTTP messages.

I'm okay with doing our own socket programming, especially given the difficulty involved in doing Rust ```async``` blocks necessary for hyper/tokio.  But I really don't want us to do our own protocol parsing; I've written HTTP parsers at least 3 times, and it was always a nightmare.  

So if we're gonna use 3rd party utilities then I'm good with this.

**Comment by sergey-shandar** [comment]
> > > One high level question: why are we implementing HTTP instead of simply using an existing library (e.g. hyper)?
> > 
> > 
> > We would like to use the code in test and easily mock any IO. However, we may use some utility functions from third-party libraries, for example encode/decode HTTP messages.
> 
> I'm okay with doing our own socket programming, especially given the difficulty involved in doing Rust `async` blocks necessary for hyper/tokio. But I really don't want us to do our own protocol parsing; I've written HTTP parsers at least 3 times, and it was always a nightmare.
> 
> So if we're gonna use 3rd party utilities then I'm good with this.

I will open an issue for this.

**Pull request merged**